### PR TITLE
ENTMQSTPR-14: Examples for Proxy / Record Encryption

### PR DESCRIPTION
**Pull request by k-wall**
We want the customer to try AMQ Streams Tech Preview with Record Encryption as easily as possible.

This PR contributes examples that let the customer do that in an OpenShift environment.

There are two examples:

* load-balancer which deploys proxy exposed by an External Load Balancer.
* cluster-it which deploys proxy exposed within the cluster by a Cluster-IP.

In both case, the proxied Kafka Cluster is stood up on the same OpenShift Cluster.  There are instructions that let the customer either deploy a development HashiCorp Vault, or use a suitable one they already have.

### CODE REVIEW

**Comment by k-wall** [comment]
@SamBarker FYI - not ready for a review... It is a car crash ATM!

File: examples/proxy/envelope-encryption/proxy_exposed_by_loadbalancer/README.md
Comment (k-wall) [reply examples/proxy/envelope-encryption/proxy_exposed_by_loadbalancer/README.md]: Glad you found the file!
Sorry for the hunt.

**Comment by k-wall** [comment]
@SamBarker It's starting to take a bit for shape now.
The instructions for using an existing vault instate or installing your own are hopefully close.
The example proxy_exposed_by_loadbalancer seems to be working on OpenShift.
I still need to put together the other examples.  Tomorrow.

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (k-wall) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:13]:
```suggestion
### Prerequisites
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (k-wall) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:8]: Need to talk about the HashiCorp version (1.15)

File: examples/proxy/envelope-encryption/base/proxy/proxy-deployment.yaml
Comment (k-wall) [inline examples/proxy/envelope-encryption/base/proxy/proxy-deployment.yaml:26]: Use downstream

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (k-wall) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:44]: FYI If we get https://github.com/kroxylicious/kroxylicious/pull/964 done, these `sed` steps will go away.  Instead, we'll create a config map here.  That CM will be mounted into the pod, and an `envFrom` to pull the CM value into the pod's environment.  Then the interpolation will put the value into the file.

File: examples/proxy/envelope-encryption/base/kafka/kafka-persistent.yaml
Comment (franvila) [inline examples/proxy/envelope-encryption/base/kafka/kafka-persistent.yaml:9]: Is that right? Do we want `3.6.0` or `3.6.1`?

**Review by franvila — APPROVED** [review]
LGTM. Small nit with Kafka Version used

File: examples/proxy/envelope-encryption/base/kafka/kafka-persistent.yaml
Comment (k-wall) [reply examples/proxy/envelope-encryption/base/kafka/kafka-persistent.yaml:9]: It'll be updated to 3.7 before it goes out.  I notice that the examples still reference `3.6.0`.

https://github.com/k-wall/amqstreams-1-openshift-image/blob/ENTMQSTPR-14/examples/kafka/kafka-ephemeral-single.yaml

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (k-wall) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:77]: We should refer to the Helm chart by <major>.<minor> so the customer picks up any bug fixes to the chart itself.

File: examples/proxy/envelope-encryption/vault/helm-dev-values.yaml
Comment (k-wall) [inline examples/proxy/envelope-encryption/vault/helm-dev-values.yaml:15]: Get rid of the comment out code.  Btw.  This doesn't work - it races.

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:8]:
```suggestion
You may either use an existing instance of HashiCorp Vault you have available in your organisation, or deploy a development instance.
```

Comment (PaulRMellor) [reply examples/proxy/envelope-encryption/PREPARE_KMS.md:8]: Can't we use a production instance?

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:9]:
```suggestion
Deploying HashiCorp Vault to your OpenShift cluster provides an adequate environment to try out the filter.
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:17]:
```suggestion
You must have an HashCorp Vault instance running on-premise, or access to HashiCorp HCP Vault or Enterprise.  
The free tier of HCP Vault Secrets is *not*  sufficient.

```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:21]:
```suggestion
* You must have administrative access to the Vault instance or Vault namespace.
* You must know the address of the Vault server, and if Vault Enterprise, the name of the Vault namespace(s).
* Vault CLI is installed.
* GNU sed is available.
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:23]:
```suggestion
NOTE: These examples are prepared against HashiCorp Vault 1.15.
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:48]:
```suggestion
6. To complete the process, follow the steps to [Create the Vault Policy and Filter Vault Token](#create-the-vault-policy-and-filter-vault-token)
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:53]:
```suggestion
This option deploys a development instance of HashiCorp Vault with ephemeral storage.  This is sufficient to help you evaluate the
filter, but is *not* suitable for production or controlled test environments.
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:55]:
```suggestion
It is assumed that you'll be deploying the HashiCorp Vault feature to the same OpenShift cluster where you'll be evaluating AMQ Stream Proxy.
```

Comment (PaulRMellor) [reply examples/proxy/envelope-encryption/PREPARE_KMS.md:55]: In Streams the _house style_ is we don't capitalize things like _cluster_ only proper nouns like Openshift. I don't know if we want to go along with that?  I won't continue suggesting in this review.

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:63]:
```suggestion
### Deploying HashiCorp Vault for Development
```

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/PREPARE_KMS.md:110]:
```suggestion
After completing the task, if you have deployed Vault, remove it as follows:
```

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:5]:
```suggestion
to decrypt the messages before they are returned to consuming applications.
```

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:10]: would people know what key material is?  I didn't!

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:11]:
```suggestion
Proxy.  The KMS is external and must be provided by the system administrator or deployer.  For this Tech Preview release, only HashiCorp Vault is supported.
```

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:13]:
```suggestion
To be able to use the filter, you must prepare a HashiCorp Vault instance:
```

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:16]:
```suggestion
* Otherwise, there are [deployment instructions](./PREPARE_KMS.md#deploying-hashicorp-for-development) to install an *ephemeral development* instance to your OpenShift Cluster.
```

File: examples/proxy/envelope-encryption/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/README.md:18]:
```suggestion
Once you have a Vault instance prepared, you can proceed to deploy one of the examples.  
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:3]:
```suggestion
In this example, an instance of Apache Kafka is deployed using AMQ Streams.  The instance is proxied using
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:13]:
```suggestion
* OpenShift CLI (`oc`)
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:14]: cluster-wide?

Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:14]:
```suggestion
* AMQ Streams Operator installed namespace wide
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:13]:
```suggestion
* OpenShift CLI (`oc`)
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:14]: cluster-wide?

Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:14]:
```suggestion
* AMQ Streams Operator is installed namespace wide
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:37]:
```suggestion
4. Consume messages *direct* from the Kafka Cluster, showing that they are encrypted:
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:41]:
```suggestion
5. Consume messages from the *proxy* showing they get decrypted automatically:
```

File: examples/proxy/envelope-encryption/cluster-ip/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/cluster-ip/README.md:18]:
```suggestion
1. Deploy the Example:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:19]:
```suggestion
1. Deploy the Example:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:23]:
```suggestion
2. Get the external address of the proxy service:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:27]:
```suggestion
3. Now update the `brokerAddressPattern:` to match the `LOAD_BALANCER_ADDRESS`:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:31]:
```suggestion
4. Reapply and bounce:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:48]:
```suggestion
4. Produce some messages to the topic:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:52]:
```suggestion
5. Consume messages direct from the Kafka Cluster, showing that they are encrypted:
```

File: examples/proxy/envelope-encryption/load-balancer/README.md
Comment (PaulRMellor) [inline examples/proxy/envelope-encryption/load-balancer/README.md:56]:
```suggestion
6. Consume messages from the proxy showing they are decrypted:
```

**Review by PaulRMellor — COMMENTED** [review]
Thanks Keith. I've made a few minor suggestions, some relating to consistency as I read through.

File: examples/proxy/envelope-encryption/PREPARE_KMS.md
Comment (k-wall) [reply examples/proxy/envelope-encryption/PREPARE_KMS.md]: Technically speaking, any Vault instance with the transit engine enable will function just fine.  Should the customer be using their production Vault instance to experiment with a Tech Preview feature? Probably no.   Do we need to spell that out?

**Comment by k-wall** [comment]
Thanks @PaulRMellor, I incorporated your suggestions into the document.

With regard to the question what we we call the feature.   In the code we've been used EnvelopeEncryption (that's the name of the encryption technique we've applied).  It is also the name that appears in the configuration.    We've used the term 'Topic Encryption' and 'Encryption At Rest'. Maybe we should run the question past the BU?

**Comment by k-wall** [comment]
@PaulRMellor I've updated the example to use the term Record Encryption.

**Comment by k-wall** [comment]
@PaulRMellor. @SamBarker . I've updated the examples to enable metrics.  The `oc apply -k` will apply the pod-monitor resource as the deployment is made.  So if the user has configured AMQ Streams for metrics (using the normal instructions), Kroxylicious metrics will be available too.

There's no Proxy Grafana dashboard - I think we consider this out of scope.

**Comment by SamBarker** [comment]
> There's no Proxy Grafana dashboard - I think we consider this out of scope.
Agree for tech preview. 

As an aside can you tell why there are two `API_VERSIONS` series? Looks like its request vs response based on the flowing tags. We should probably include some documentation on the tags we add to the metrics. Given we have distinct Request and Response Interfaces should we also have distinct metrics? 

cc: @robobario

**Review by PaulRMellor — APPROVED** [review]

**Comment by k-wall** [comment]
> > There's no Proxy Grafana dashboard - I think we consider this out of scope.
> > Agree for tech preview.
> 
> As an aside can you tell why there are two `API_VERSIONS` series? 

`io.kroxylicious.proxy.internal.SaslDecodePredicate#shouldDecodeRequest` insists that the API_VERSION request from the client is decoded.  AFAIK we're not actually relying on that right now (it is done for SASL offload IIRC).  For the response, it is ApiVersionsIntersectFilter that is insisting the response is decoded.  That explains the two rows.

> Looks like its request vs response based on the flowing tags. We should probably include some documentation on the tags we add to the metrics. 

Yes, it'd be good to get that in.

> Given we have distinct Request and Response Interfaces should we also have distinct metrics?

You've lost me.  Don't we have that already?  What we don't have is metrics showing tracking what ultimately gets sent upstream, or ultimately sent downstream.  If a filter changes the request/response size, or sends a request out of band, or even drops a request, the metrics are blind to that.

> 
> cc: @robobario

**Comment by k-wall** [comment]
@SamBarker if you are happy with the content, I'd like to merge this PR.

**Comment by SamBarker** [comment]
> You've lost me. Don't we have that already? What we don't have is metrics showing tracking what ultimately gets sent upstream, or ultimately sent downstream. If a filter changes the request/response size, or sends a request out of band, or even drops a request, the metrics are blind to that.

No, I think you've followed me. I was confused by the duplicate AP_VERSIONS or METADATA series. Half way through writing the comment I realised that it was request and response for each. So I added the bit about documenting the flowing tag and then didn't properly re-phrase the rest. What I was trying to say is given we have distinct interfaces for requests and responses in the API doesn't the same logical distinction make sense in the metrics (if its sensible in one part of the public API it should be a high bar for changing else where). 


I've been trying to run the content locally before approving but have stalled on issues with my CRC cluster setup (its failing to pull vault). I don't think I'm going to get to it today but should be able to try again with minikube to unblock this in a day or two.

**Comment by robobario** [comment]
The metrics we produce on a per API basis are:

```
kroxylicious_payload_size_bytes_count{ApiKey="PRODUCE",ApiVersion="5",flowing="downstream",} 5.0
kroxylicious_payload_size_bytes_count{ApiKey="PRODUCE",ApiVersion="5",flowing="upstream",} 5.0
kroxylicious_payload_size_bytes_max{ApiKey="PRODUCE",ApiVersion="5",flowing="downstream",} 54.0
kroxylicious_payload_size_bytes_max{ApiKey="PRODUCE",ApiVersion="5",flowing="upstream",} 123.0
kroxylicious_payload_size_bytes_sum{ApiKey="PRODUCE",ApiVersion="5",flowing="downstream",} 270.0
kroxylicious_payload_size_bytes_sum{ApiKey="PRODUCE",ApiVersion="5",flowing="upstream",} 615.0
```

with labels:

- ApiKey: the Kafka protocol API, the name from `org.apache.kafka.common.protocol.ApiKeys` enum.
- ApiVersion: the version of the api (each ApiKey can have multiple versions, client A sends Produce at v5 and client B sends Produce at v6)
- flowing: "upstream" || "downstream". "upstream" is measuring messages arriving at the Proxy from the client in the `KafkaRequestDecoder`, so flowing upstream towards the proxied broker. "downstream" is measuring messages received from the proxied broker in the `KafkaResponseDecoder`, flowing towards the client.

We could change the tagging to be oriented around Request/Response naming. Or we could construct the name like `message: ${camelcase(ApiKey.name)}Request|Response`. We are working with the ApiKey off the wire at that point, the idea of request/response is contextual "I'm a server so it's a request" or "I'm receiving something from the upstream, so it's a response". We could code gen this in the api module to centralise it, something like `ApiKeyNames.request(ApiKeys key)` and `ApiKeyNames.response(ApiKeys key)`.

**Comment by k-wall** [comment]
> edit: thinking about it, tags might be useful for visualising an ApiKey's request/response so maybe tagging is better

I think the having the separate label for the ApiKey is in line with Prometheus best practice.  However I did raise https://github.com/kroxylicious/kroxylicious/issues/1060

File: examples/proxy/envelope-encryption/base/proxy/proxy-deployment.yaml
Comment (MikeEdgar) [reply examples/proxy/envelope-encryption/base/proxy/proxy-deployment.yaml]: @k-wall is the plan to simply keep this manually up-to-date for each downstream build?

**Comment by k-wall** [comment]
@SamBarker I know you started reviewing this PR, have you done yet?

File: examples/proxy/record-encryption/README.md
Comment (SamBarker) [inline examples/proxy/record-encryption/README.md:3]: nit? the filter is the `RecordEncryption` so `The Encryption filter` feels a bit of an ambiguous reference.

File: examples/proxy/record-encryption/README.md
Comment (SamBarker) [inline examples/proxy/record-encryption/README.md:11]:
```suggestion
The filter relies on a Key Management System (KMS). The role of the KMS is to provide cryptographic functions and act as a repository of key-material. The KMS is *not* part of AMQ Streams.  It is external must be provided by the deployer of the system.  For this Tech Preview release, only HashiCorp Vault is supported.
```

File: examples/proxy/record-encryption/base/kafka/kafka-persistent.yaml
Comment (SamBarker) [inline examples/proxy/record-encryption/base/kafka/kafka-persistent.yaml:9]: Should this be 3.7.0 as part of the upcoming release?

**Review by SamBarker — APPROVED** [review]
I've been unable to get my cluster to pull the vault image (I have no reason to believe its anything to do with this PR) but rather than hold this PR up any further I've reviewed with out verifying.

File: examples/proxy/record-encryption/PREPARE_KMS.md
Comment (k-wall) [reply examples/proxy/record-encryption/PREPARE_KMS.md:39]: Done.

**Pull request merged**
